src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs

```diff
@@ -453,10 +517,58 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
                         _cteToLimit.Add(_tableExpressionCounter);
                     }
 
+                    // Update target reference cte dictionary
+                    var curLimitCte = TableExpressionName(_tableExpressionCounter + 1);
+
+                    // Add current cte limit to the dictionary
+                    if (includeExpression.Reversed)
+                    {
+                        AddIncludeLimitCte(includeExpression.SourceResourceType, curLimitCte);
+                    }
+                    else
+                    {
+                        // Not reversed and a specific target type is provided as the 3rd part of include value
+                        if (includeExpression.TargetResourceType != null)
+                        {
+                            AddIncludeLimitCte(includeExpression.TargetResourceType, curLimitCte);
+                        }
+                        else if (includeExpression.ReferenceSearchParameter != null)
+                        {
+                            includeExpression.ReferenceSearchParameter.TargetResourceTypes?.ToList().ForEach(t => AddIncludeLimitCte(t, curLimitCte));
+                        }
+                    }
+
+                    // Handle Multiple Results sets to include from
+                    if (_includeFromCteIds?.Count > 1 && _curFromCteIndex >= 0 && _curFromCteIndex < _includeFromCteIds.Count - 1)
```

Comment:
I feel like we could take complexity out of this class by figuring this stuff out in a visitor pass ahead of this. Each CTE we generate would be based on a TableExpression in the SqlRootExpression, and we could figure out, in the case of iterate expression, which TableExpression is is meant to build off of. Admittedly, taking direct references could be tricky as the trees are immutable and changing them requires replacing them. (We might want to use IDs and a symbol table).

Reply:
If I understand you correctly, then you mean building the dictionary of ctes to include from before all the visits, instead of building it on the fly. This is something I raised when we discussed recursive queries.

This will require a major  change in all the VisitSqlRoot method which I don;t think should be out of scope for this PR

Reply:
Yeah, let's leave it as it is now. I do feel like the overall complexity in this class is getting higher than I would like, so I'd like to do some refactoring in the future.